### DISCUSSION TITLE
[Proposal] Conditional out keyword

### DISCUSSION THREAD

**Discussion by Per6** [dotnet/csharplang#8544] [Language Ideas] [answered]
### Summary
I propose to add a trueout and falseout keyword, that are out keywords, with the addition, that the compiler knows, that all trueout's exist if the function returned true and all falseout's exist if the function returned false. This will make out safer and easier to use.
(names and/or syntax should probably be changed)
### Examples
To use it simply define a function, that has to return a boolean with one or more trueout or falseout parameters.
the `return` signals a return of true and the use of all the trueout parameters, while the `return!` signals for false and falseout parameters.
```
bool Example(trueout int num1, falseout float num2, falseout float? num3) {
    if(someCondition)
        return 5; # <- this returns true and sets num1 to be 5
    return! (0.5f, null); # <- this returns false and sets num2 to be 0.5f and num3 to be null
}
```
Then you can use it like this:
```
if(Example(trueout int num1, falseout float num2, falseout float? num3))
    OtherFunction1(num1);
else
    OtherFunction2(num2, num3);
```
It should throw a compile error in these situations, which would make it safer to use than simply out:
```
if(Example(trueout int num1, falseout float num2, falseout float? num3))
     OtherFunction3(num2); <- Err: Use of unassigned local variable 'num2'
else
     OtherFunction4(num1); <-Err:  Use of unassigned local variable 'num1'
OtherFunction5(num3); <- Err: Use of unassigned local variable 'num3'
```
```
int ErrorExample(trueout int num1, trueout int num2) { # <- Err: does not return a bool despite trueout and/or falseout existing in the parameters
    if(someCondition)
        return! (5, 0.5f); # <- Err: not enough falseout parameters to capture the return values
    return 0.5f; # <- Err: wrong type: expected int but returns float (Also Err: not enough return values to fill all parameters)
}
```
Each trueout parameter should effectively be exactly the same as this:
```
if(someNullableValue is {} createdVariable)
```
Each falseout parameter should effectively be exactly the same as this:
```
if(someNullableValue is not {} createdVariable)
```

**Comment by HaloFour** [comment]
What are the use cases for a feature like this?  Seems like something you could accomplish with an analyzer that does some flow analysis and a custom attribute on the `out` parameter.

**Comment by Per6** [comment]
My use case is returning either a value or returning a reason for failure to be handeled outside of the function. There are also a lot of places, where i could replace returning an optional value and then pattern matching it to a value, to make it look nicer without compromising safety.
I have never created an analyzer so i'll need to look into that.

**Comment by huoyaoyuan** [comment, marked as answer]
Functionally the same as #8530 . This definitely won't be considered as a keyword.
